Build failure: `pkgsStatic.dnsperf`

## Steps To Reproduce

Steps to reproduce the behavior:

1. build `pkgsStatic.dnsperf`

## Build log

<!-- insert build log in code block in collapsable section -->

<details>

<summary>Build Log</summary>

```
$ nix-build '<nixpkgs>' --impure --attr "pkgsStatic.dnsperf" --cores "$(($(nproc)+1))" --max-jobs "$(($(nproc)+1))" --log-format bar-with-logs
these 4 derivations will be built:
  /nix/store/2i0z44a1r58y8bkkzxm89r13v7jli8q6-ck-static-x86_64-unknown-linux-musl-0.7.2.drv
  /nix/store/lgzydnl3kc5qz43062vf3c2grq8pl540-dns-root-data-static-x86_64-unknown-linux-musl-2024-06-20.drv
  /nix/store/sna0jd3ikxakyqvhvq4m2c0d5fdbn7bv-ldns-static-x86_64-unknown-linux-musl-1.8.4.drv
  /nix/store/ix6jwg29yh16d9abg9skwz3ccxpjmjcp-dnsperf-static-x86_64-unknown-linux-musl-2.14.0.drv
---snipped--
dnsperf-static-x86_64-unknown-linux-musl> x86_64-unknown-linux-musl-gcc -DHAVE_CONFIG_H -I.    -I. -I.. -pthread -I/nix/store/ykfg6sbazxq50frh5aigrwgxbv7cjvr9-openssl-static-x86_64-unknown-linux-musl-3.3.2-dev/include -I/nix/store/ykfg6sbazxq50frh5aigrwgxbv7cjvr9-openssl-static-x86_64-unknown-linux-musl-3.3.2-dev/include -I/nix/store/lh2yfp3lzdl1k40657pk5zfnnlv3hsy1-ldns-static-x86_64-unknown-linux-musl-1.8.4-dev/include -g -O2 -D__unknown__ -DCK_USE_CC_BUILTINS=1 -I/nix/store/84nmik09z723inm2ylb0d7bdp23pyzsp-ck-static-x86_64-unknown-linux-musl-0.7.2/include  -c -o strerror.o strerror.c
dnsperf-static-x86_64-unknown-linux-musl> strerror.c: In function 'perf_strerror_r':
dnsperf-static-x86_64-unknown-linux-musl> strerror.c:35:12: error: returning 'int' from a function with return type 'const char *' makes pointer from integer without a cast []
dnsperf-static-x86_64-unknown-linux-musl>    35 |     return strerror_r(errnum, str, len);
dnsperf-static-x86_64-unknown-linux-musl>       |            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
dnsperf-static-x86_64-unknown-linux-musl> make[3]: *** [Makefile:669: strerror.o] Error 1
dnsperf-static-x86_64-unknown-linux-musl> make[3]: Leaving directory '/build/source/src'
dnsperf-static-x86_64-unknown-linux-musl> make[2]: *** [Makefile:741: all-recursive] Error 1
dnsperf-static-x86_64-unknown-linux-musl> make[2]: Leaving directory '/build/source/src'
dnsperf-static-x86_64-unknown-linux-musl> make[1]: *** [Makefile:481: all] Error 2
dnsperf-static-x86_64-unknown-linux-musl> make[1]: Leaving directory '/build/source/src'
dnsperf-static-x86_64-unknown-linux-musl> make: *** [Makefile:475: all-recursive] Error 1
error: builder for '/nix/store/ix6jwg29yh16d9abg9skwz3ccxpjmjcp-dnsperf-static-x86_64-unknown-linux-musl-2.14.0.drv' failed with exit code 2
```

</details>

## Metadata
```bash
$ nix-shell -p nix-info --run "nix-info -m"
bash: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8): No such file or directory
/nix/store/5mh7kaj2fyv8mk4sfq1brwxgc02884wi-bash-5.2p37/bin/bash: warning: setlocale: LC_ALL: cannot change locale (C.UTF-8): No such file or directory
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.0-25-amd64, Ubuntu, 24.04.1 LTS (Noble Numbat), nobuild`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.25.3`
 - nixpkgs: `/nix/store/qx8njj2ikv1kizbf5pbfbza9f8k9vq50-source`
```

## Notify maintainers
@jtojnar

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/by-name/dn/dnsperf/package.nix
+++ b/pkgs/by-name/dn/dnsperf/package.nix
@@ -3,6 +3,7 @@
   stdenv,
   autoreconfHook,
   fetchFromGitHub,
+  fetchpatch,
   ldns,
   libck,
   nghttp2,
@@ -31,6 +32,16 @@ stdenv.mkDerivation rec {
     libck
     nghttp2
     openssl
+  ];
+
+  patches = lib.optionals stdenv.hostPlatform.isMusl [
+    # dnsperf doesn't have support for musl (https://github.com/DNS-OARC/dnsperf/issues/265)
+    # and strerror_r returns int on non-glibc: https://github.com/NixOS/nixpkgs/issues/370498
+    # TODO: remove if better non-glibc detection is ever upstreamed
+    (fetchpatch {
+      url = "https://gitlab.alpinelinux.org/alpine/aports/-/raw/5bd92b8f86a0bf15dddf8fa180adf14344d6cc15/testing/dnsperf/musl-perf_strerror_r.patch";
+      hash = "sha256-yTJHXkti/xSklmVfAV45lEsOiHy7oL1phImNTNtcPkM=";
+    })
   ];
 
   doCheck = true;

